---
title: "CFSpIT Microbiota Analysis R Code"
author: "Rebecca Weiser"
date: "2 September 2021"
output: html_document
---


##Analysis overview

This R code was used to perform analysis of the decontaminated of sequence reads. The decontamination was performed in Microsoft Excel and the resulting OTU table and metadata table are available in the Supplementary Excel file.

The scripts describe:

1. **Phyloseq for read processing and subsampling** (Supplementary Table S1) **and generation of an OTU table consolidated to genus level**
2. **Alpha diversity analysis** (Figure 1A)
3. **Beta diversity analysis** (Figure 1B, Figure 2, Figure 3, Supplementary Figure S2, Table 2, Supplementary Table S2)
4. **Generation of stacked bar charts to visualise microbiota profiles** (Figure 4, Supplementary Figure S3)
5. **Checking concordance between mock community and DNA extraction blank bacterial diversity profiles** (Supplementary methods and results)

Other analysis was performed in Microsoft Excel, SPSS and MedCalc software.

Other tables/figures were generated using Microsoft Excel (Figure 5A, Supplementary Figure S1) and Venny (Figure 5B). Microsoft PowerPoint was used to produce final versions of figures.


##Phyloseq for read processing and subsampling

Import data into phyloseq and create a phyloseq object

```{r}
library(phyloseq)

#Set working directory
setwd("E:/CLIMB_data/Volume2/CF-SpiT_merge/trim_galore_output/trimmed/no_subsample/Beky_decon/Final_decon/R_markdown_files")

#Define files for import
#sharedfile is the decontaminated OTU table saved in the format of the mothur shared file
#Columns are label (0.03), Group (sample name), numOtus (1743) and then indivdual Otus 
sharedfile="Beky_decon_shared.shared"
#taxfile is the mothur taxonomy file
taxfile="CFSPiT.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.opti_mcc.unique_list.0.03.cons.taxonomy"
#map file is the file containing the metadata for each sample
#Columns are Group (sample name), Patient code and Sample type
mapfile="Beky_decon_env_file.csv"

#Import files and merge to create a phyloseq object
mothur_data <- import_mothur(mothur_shared_file = sharedfile, mothur_constaxonomy_file = taxfile)
map <- read.csv(mapfile)
map <- sample_data(map)
rownames(map) <- map$Group
moth_merge <- merge_phyloseq(mothur_data, map)

#View phyloseq object
moth_merge
```

Read processing by keeping OTUs occuring >=10 times across the dataset (OTUs occuring less than 10 times are removed)

```{r}
moth_merge_lessthan10 = prune_taxa(taxa_sums(moth_merge)>=10, moth_merge)

#View phyloseq object
moth_merge_lessthan10
```

Subsample each sample to 1000 reads

```{r, warning=FALSE, error=FALSE, message=FALSE}
#Use rngseed(123) to perform repeatable random subsampling
moth_merge_lessthan10_subsample1000 <- rarefy_even_depth(moth_merge_lessthan10, sample.size = 1000, rngseed = 123, replace = FALSE)

#View phyloseq object
#Four samples have been lost after subsampling (CF102BAL3 CF109BAL3 CF169BAL2 and CF178BAL2)
moth_merge_lessthan10_subsample1000
```

Calculate coverage before and after subsampling

```{r, warning=FALSE, error=FALSE, message=FALSE}
library(metagMisc)
library(iNEXT)

#Original phyloseq object
coverage_original<-phyloseq_coverage(moth_merge)
colnames(coverage_original)[colnames(coverage_original) == "SampleCoverage"] <- "SampleCoverage_original"

#After filtering <10 reads then subsampling to 1000 (lose 4 samples)
coverage_subsampled<-phyloseq_coverage(moth_merge_lessthan10_subsample1000)
colnames(coverage_subsampled)[colnames(coverage_subsampled) == "SampleCoverage"] <- "SampleCoverage_subsampled"

#Save files to compare coverage before and after read processing
write.csv(coverage_original, file = "decontam_coverage_original.csv")
write.csv(coverage_subsampled, file = "decontam_coverage_subsampled.csv")
```


##Combining OTU and taxonomy files to generate a single OTU table for further analysis

```{r, warning=FALSE, error=FALSE, message=FALSE}
#Generate a separate OTU table from the phyloseq object and save the OTU table as a matrix in R
moth_merge_lessthan10_subsample1000_OTU = as(otu_table(moth_merge_lessthan10_subsample1000), "matrix")
#Coerce this matrix to a dataframe
moth_merge_lessthan10_subsample1000_OTU_DF = as.data.frame(moth_merge_lessthan10_subsample1000_OTU) 

#Write to .csv, then open the file in Excel, change top left hand cell to OTU_ID and save as .txt
write.csv(moth_merge_lessthan10_subsample1000_OTU_DF, file="CFSpIT_Beky_decon_lessthan10_subsample1000.csv")

#Open the mothur taxonomy file (CFSPiT.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.opti_mcc.unique_list.0.03.cons.taxonomy), delete middle 'Size' column, change top left hand cell to OTU_ID and save as .txt (Beky_decon_BAL123_IS_taxfile.txt)

#Read tables into R and merge based on OTU_ID column heading
otu <- read.table("CFSpIT_Beky_decon_lessthan10_subsample1000.txt", header = T, check.names = F) 
tax	<- read.table("Beky_decon_BAL123_IS_taxfile.txt", header = T, check.names = F)
merged_file	<- merge(otu, tax, by.x	= c("OTU_ID"), by.y=c("OTU_ID"))
View(merged_file)

#Re-order the columns in merged_file to a more useful order with column 1 as OTU_ID and column 2 as taxonomy
library(dplyr)
merged_file_taxon_first <- merged_file %>% select(OTU_ID, Taxonomy, everything())
View(merged_file_taxon_first)

#Write to .txt
write.table(merged_file_taxon_first, "Beky_decon_BAL123_IS_4samples_excluded_OTU_table.txt",	sep	= '\t',	col.names =	TRUE, row.names	= FALSE)

#Open this file in Excel, consolidate the reads for each sample by the genus rank then save as a .txt. The resulting OTU table will have  column 1 as genus (no column header) and the other columns as samples with sample name headers.

```


##Analysis of the OTU table consolidated to genus level (contains 116 samples)

###Alpha diversity analysis

This analysis requires OTU tables and accompanying metadata files which contain information about the samples.

The metadata file was created in Excel with Sample ID as the first column and other information (e.g. patient code, sample type, patient age) in the other columns and saved as a .csv. The file has the four samples excluded after subsampling removed and has the sample IDs in the same order as the OTU table.

We will firstly read in the files and check that the sample IDs are in the same order in each file

```{r, warning=FALSE, error=FALSE, message=FALSE}
#Read in OTU table and metadata table
bacterial_data<-read.table("Beky_decon_BAL123_IS_4samples_excluded_OTU_table_R.txt", header=T)
meta_table<-read.csv("Beky_decon_4samples_excluded_env.csv", header = T)

#check that the sample names are in the same order in the OTU table (across) and the metadata file (down)
#Save OTU table sample ID as an object (currently column names)
Sample_ID_OTU<-colnames(bacterial_data)

#Save env file sample ID (currently in SAMPLE column)
Sample_ID_env<-as.character(meta_table$SAMPLE)

#check they are the same order
all.equal(Sample_ID_OTU, Sample_ID_env)
```

Then we can calculate the alpha diversity of the samples (Shannon, Simpson and Richness) and add these values into the metadata file for each sample.

```{r, warning=FALSE, error=FALSE, message=FALSE}
#ALPHA DIVERSITY

#transpose the OTU table
data_transpose<-t(bacterial_data)   
matrix_data<-data.matrix(data_transpose) 

#Calculate diversity metrics
library(vegan)

shannon.diversity <- diversity(matrix_data, "shannon")
#Make data frame
shannon.diversity <- as.data.frame(shannon.diversity) 
#Make SAMPLE into a column
shannon.diversity<-cbind(SAMPLE=row.names(shannon.diversity),shannon.diversity)
#Remove rownames
row.names(shannon.diversity) <- NULL

simpson.diversity <- diversity(matrix_data, "simpson") 
simpson.diversity <- as.data.frame(simpson.diversity)
simpson.diversity <- cbind(SAMPLE=row.names(simpson.diversity), simpson.diversity)
row.names(simpson.diversity) <- NULL

richness<-specnumber(matrix_data)
richness <- as.data.frame(richness)
richness <- cbind(SAMPLE=row.names(richness), richness)
row.names(richness) <- NULL

#Add to metadata file, have to do multiple merges (R 'merge' can only merge 2 dataframes at a time)
#Used merge as it makes sure everything is in the same order (uses common column SAMPLE)
meta_table<-merge(meta_table, shannon.diversity, by = "SAMPLE")
meta_table<-merge(meta_table, simpson.diversity, by = "SAMPLE")
meta_table<-merge(meta_table, richness, by = "SAMPLE")

#Check metadata table, look at alpha diversity values
View(meta_table)
```

We will focus on the Shannon diversity values and want to compare the diversity between different sample types. For this we will remove the sets with missing values (CF102, CF109, CF169 and CF178), resulting in 78 samples.

```{r, warning=FALSE, error=FALSE, message=FALSE,fig.dim=c(6,6)}
#Subset the metadata table
Subset_meta_table_26sets<- subset(meta_table, Patient!="CF102" & Patient!="CF109" & Patient!="CF169" & Patient!="CF178")

#Draw boxplots using ggplot to compare the diversity between sample types
#Figure 1A
library(ggplot2)

plot1 <- ggplot(data=Subset_meta_table_26sets, aes(x=`Type`, y=shannon.diversity)) + geom_boxplot(aes(fill=`Type`)) + ylim(0,3) + theme_bw() +   theme(panel.background = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + ylab("Shannon diversity") + xlab("") + theme(axis.ticks.length=unit(.25, "cm")) + theme(text = element_text(size = 20))

plot1 + scale_fill_manual(values=c("#1F78B4", "#33A02C", "#ff7f00", "black")) + theme(legend.position = "none")  + theme(axis.text.x = element_text(vjust = -2))
```

Statistical testing to determine if there are significant differences in Shannon diversity between sample types

```{r, warning=FALSE, error=FALSE, message=FALSE}
#Check data distribution of shannon diversity for all samples
hist(Subset_meta_table_26sets$shannon.diversity, main = "BAL1, BAL2, BAL3 and IS samples", xlab = "Shannon diversity index", ylab = "Frequency")
shapiro.test(Subset_meta_table_26sets$shannon.diversity)

#Data distribution does not look normal and have significant Shapiro-Wilk result
#Check data distribution for each sample type
by(Subset_meta_table_26sets$shannon.diversity, Subset_meta_table_26sets$Type, shapiro.test) 
#Significant results for BAL1, BAL2 and IS for Shapiro-Wilk test so not normally distributed

bartlett.test (Subset_meta_table_26sets$shannon.diversity ~  Subset_meta_table_26sets$Type) 
#Significant result for Bartlett test suggests unequal variances

#Need to run a non-parametric test for multiple groups that accounts for repeated measures
#Use a Friedman Test

library(PMCMR)

friedman.test(Subset_meta_table_26sets$shannon.diversity, groups = Subset_meta_table_26sets$Type, blocks = Subset_meta_table_26sets$Patient)

#Signficant Friedman Test result so run post hoc tests
posthoc.friedman.conover.test(y = Subset_meta_table_26sets$shannon.diversity, groups = Subset_meta_table_26sets$Type, blocks = Subset_meta_table_26sets$Patient, p.adjust.method="bonferroni")

#BAL1-IS, BAL2-IS and BAL3-IS are significantly different (BALs aren't)
```


###Beta diversity analysis 

####NMDS ordination looking at different sample types as groups

We will perform this using the 26 complete sample sets (BAL1, BAL2, BAL3 and IS). It uses an edited OTU table with CF102, CF109, CF169 and CF178 samples removed (resulting in 78 samples), and the corresponding metadata table generated above.


```{r, warning=FALSE, error=FALSE, message=FALSE,fig.dim=c(6,6)}
#Read in OTU table that has been edited to contain only the 26 complete BAL123_IS sets
bacterial_data<-read.table("Beky_decon_BAL123_IS_26sets_OTU_table_R.txt", header=T) 
transposed_bacterial_data<-t(bacterial_data) 

#Change the Enterobacteriaceae_unclassified column name to Enterobacteriaceae so it can be displayed nicely on a figure
colnames(transposed_bacterial_data)[colnames(transposed_bacterial_data) == 'Enterobacteriaceae_unclassified'] <- 'Enterobacteriaceae'

sol<-metaMDS(transposed_bacterial_data,distance = "bray", k = 2, trymax = 50)

sol
#Stress value is high but less than 0.2 so ok

#Before adding in metadata, check that the sample names are in the same order in the OTU table (across) and the metadata file (down)
Sample_ID_OTU<-colnames(bacterial_data)
Sample_ID_env<-as.character(Subset_meta_table_26sets$SAMPLE)
all.equal(Sample_ID_OTU, Sample_ID_env)

#Use the Subset_meta_table_26sets metadata table and convert sample type into a factor
env.sample.type<-as.factor(Subset_meta_table_26sets$Type)

#Choose the palette for plotting the NMDS and set as default palette
mypal<-c("#1F78B4", "#33A02C", "#ff7f00", "black")
palette(mypal)

#Identify the in Top7.5% genera
library(goeveg)
top_7.5_pc<-ordiselect(transposed_bacterial_data, sol, ablim = 0.075, choices = c(1, 2), method = "axes", freq = FALSE)

#Final plot including sample points, ellipses and Top7.5% genera
#Figure 1B
plot(sol, display="sites", type="n")
points(sol, pch=19, col=env.sample.type)
ordiellipse(sol, group=env.sample.type, show.groups = "BAL1", col="#1F78B4", kind="sd", lwd=2)
ordiellipse(sol, group=env.sample.type, show.groups = "BAL2", col="#33A02C", kind="sd", lwd=2)
ordiellipse(sol, group=env.sample.type, show.groups = "BAL3", col="#ff7f00", kind="sd", lwd=2)
ordiellipse(sol, group=env.sample.type, show.groups = "IS", col="black", kind="sd", lwd=2)
ordipointlabel(sol, display="species", select = top_7.5_pc, col="black", cex=1, add = TRUE)


```


####NMDS ordination looking at different age groups

We will perform this using all of the BAL samples (n=86) but no IS samples.

```{r, warning=FALSE, error=FALSE, message=FALSE, fig.dim=c(6,6)}
#Read in OTU table that has been edited to contain only BAL1, BAL2 and BAL3 samples
bacterial_data_BAL123<-read.table("Beky_decon_BAL123_4samples_excluded_OTU_table_R.txt", header=T)
transposed_bacterial_data<-t(bacterial_data_BAL123) 

colnames(transposed_bacterial_data)[colnames(transposed_bacterial_data) == 'Enterobacteriaceae_unclassified'] <- 'Enterobacteriaceae'

sol<-metaMDS(transposed_bacterial_data,distance = "bray", k = 2, trymax = 50)
sol
#Stress is high (0.2) so NMDS might not be a very good representation, but just ok

#subset the meta_table spreadsheet
meta_table_BAL123<-meta_table[meta_table$Type == "BAL1" | meta_table$Type == "BAL2" | meta_table$Type == "BAL3",]

#Check that all the samples are in the same order for the OTU and metadata tables
Sample_ID_OTU<-colnames(bacterial_data_BAL123)
Sample_ID_env<-as.character(meta_table_BAL123$SAMPLE)
all.equal(Sample_ID_OTU, Sample_ID_env)

#Turn age group into a factor and reorder
env.sample.age<-as.factor(meta_table_BAL123$Age_group)
env.sample.age <-factor(env.sample.age, levels=c("<6", "6to12","12to18"))

#Choose the palette for plotting the NMDS and set as default palette
mypal<-c("goldenrod1", "darkorange", "red")
palette(mypal)

#Identify the in Top7.5% genera
top_7.5_pc<-ordiselect(transposed_bacterial_data, sol, ablim = 0.075, choices = c(1, 2), method = "axes", freq = FALSE)

#Final plot including sample points, ellipses, Top7.5% genera and legend
#Supplementary Figure s2
plot(sol, display="sites", type="n")
points(sol, pch=19, col=env.sample.age)
ordiellipse(sol, group=env.sample.age, show.groups = "<6", col="goldenrod1", kind="sd", lwd=2)
ordiellipse(sol, group=env.sample.age, show.groups = "6to12", col="darkorange", kind="sd", lwd=2)
ordiellipse(sol, group=env.sample.age, show.groups = "12to18", col="red", kind="sd", lwd=2)
ordipointlabel(sol, display="species", select = top_7.5_pc, col="black", cex=1, add = TRUE)
legend("topright", c("<6","6to12","12to18"), cex=1, col=c ("goldenrod1","darkorange","red"), lwd=2)
```


###Hierarchical clustering (Bray-Curtis dissimilarity values, Ward's method)

We will use all of the BAL1, BAL2 and BAL3 samples (n=86) that were used above for the NMDS analysis with age groups


```{r, warning=FALSE, error=FALSE, message=FALSE, fig.dim=c(12,8)}
#Read in OTU table that has been edited to contain only BAL1, BAL2 and BAL3 samples
bacterial_data_BAL123<-read.table("Beky_decon_BAL123_4samples_excluded_OTU_table_R.txt", header=T)

#Get relative abundance of genera for each samples
library(funrar)
OTU_table_to_format<-bacterial_data_BAL123
OTU_total<-cbind(OTU_table_to_format, total = rowSums(OTU_table_to_format))
OTU_order<-OTU_total[order(-OTU_total$total),]
OTU_order_no_total<-subset(OTU_order, select = -c(total))                 
OTU_order_no_total_matrix<-as.matrix(OTU_order_no_total)
transposed_matrix<-t(OTU_order_no_total_matrix)
OTU_relabund<-make_relative(transposed_matrix)
transposed_relabund<-t(OTU_relabund)
percent_relabund<-(transposed_relabund)*100
View(percent_relabund)

#subset the meta_table spreadsheet
meta_table_BAL123<-meta_table[meta_table$Type == "BAL1" | meta_table$Type == "BAL2" | meta_table$Type == "BAL3",]

#Have already checked that the sample IDs are in the same order for the NMDS analysis with age groups

#Choose heatmap colour palette and save as col_rev
library(colorspace)
#Pick Basic: Sequential single hue (Black-white colour scheme) 
mypalette<-choose_palette()
#Reverse the direction of the palette (high=black, low=white)
col_rev <- rev(mypalette(25)) 

#Subset Top25 genera (have already been sorted from highest to lowest above)
Top25_rownames<-rownames(percent_relabund[1:25,])

#Heatmap of Bray-Curtis dissimilarity distances clustered using Ward's method
#Top25 genera used and annotation tracks for sample type and age included
#Figure 2 (edited in powerpoint to tidy up and display only the Top10 genera)
library(NMF)
aheatmap(percent_relabund, distfun = function(x) vegdist(x, method = "bray"), hclustfun = function(x) hclust(x, method = "ward.D2"), color = col_rev, treeheight = 150, Rowv=NA, annCol = list('Sample Type'=meta_table_BAL123$Type, 'Age group'=meta_table_BAL123$Age_group), annColors = list(c("BAL1" ="#1f78b4", "BAL2"="#33a02c","BAL3"="#ff7f00"), c("<6" ="goldenrod1", "6to12"="darkorange","12to18"="red")), subsetRow =  Top25_rownames)
```
```{r, warning=FALSE, error=FALSE, message=FALSE,fig.dim=c(12,8)}
#Check the Bray-Curtis clade cut-off value by drawing a dendrogram with a scale (no heatmap)
#Use the same data that was used for the heatmap (percent relabund Top25)
#Match up the Bray-Curtis scale with the clustering (1.5 defines the 7 clades)
Top25BAL123<-percent_relabund[1:25,]
View(Top25BAL123)
BC_BAL_pcTop25<-vegdist(t(Top25BAL123), method = "bray", binary = FALSE)
h = hclust(BC_BAL_pcTop25, method = "ward.D2")
plot(h, main="Bray-Curtis dissimilarity BAL samples", sub="", xlab ="", ylab="Bray-Curtis dissimilarity", cex=0.8, hang=-1)
```


###Bray-Curtis dissimilarity distances for BAL1, BAL2, BAL3 and IS samples

This analysis will determine the Bray-Curtis dissimilarity distances for all pairwise comparisons between BAL1, BAL2, BAL3 and IS samples

```{r, warning=FALSE, error=FALSE, message=FALSE}
#Read in OTU table with all 116 samples
bacterial_data_BAL123_IS<-read.table("Beky_decon_BAL123_IS_4samples_excluded_OTU_table_R.txt", header=T)

#transpose the OTU table
OTU_table_transposed<-t(bacterial_data_BAL123_IS)

#Calculate BC distances and make into matrix
BC_BALIS<-vegdist(OTU_table_transposed, method="bray", binary = FALSE)
matrixBALIS=data.matrix(BC_BALIS)
View(matrixBALIS)

#convert matrix to pairwise list (no 0s no duplicates)
#run function
flattenCorrMatrix <- function(cormat) {											
  ut <- upper.tri(cormat)
  data.frame(
    row = rownames(cormat)[row(cormat)[ut]],
    column = rownames(cormat)[col(cormat)[ut]],
    cor  =(cormat)[ut]
  )
}

#run function on BC matrix
pairwise_flat<-flattenCorrMatrix(matrixBALIS)
#save spreadsheet
write.csv(pairwise_flat, file = "decontam_pairwise_BAL123_IS_list.csv")
```

This spreadsheet is edited in Excel to produce: (1) a spreadsheet with only BAL1, BAL2 and BAL3 comparisons (n=3655), with columns added for Patient ID and whether the comparison is within or between patients (this will be read into R as a .csv to produce Figure 3), and (2) a spreadsheet with BAL1, BAL2, BAL3 and IS within patient comparisons (n=168), which was used to produce Table 2 and Supplementary Table S2 (no further analysis in R, all performed in Excel)

```{r, warning=FALSE, error=FALSE, message=FALSE}
#Read the BAL1, BAL2 and BAL3 comparison file (n=3655) into R
BAL123_BC<-read.csv(file="decontam_pairwise_BAL123_list_within_between_R.csv",check.names=FALSE)
colnames(BAL123_BC)

#Convert Patient column into a factor
BAL123_BC$Patient<-as.factor(BAL123_BC$Patient)
#All the within comparisons have a Patient number, all the between comparisons have 'No patient'
levels(BAL123_BC$Patient)

#Convert Within_Between column into a factor
BAL123_BC$Within_Between<-as.factor(BAL123_BC$Within_Between)
levels(BAL123_BC$Within_Between)

#Data preparation
#subset data within
data_subset_within<-subset(BAL123_BC, Within_Between=="Within")
View(data_subset_within)

#order patient number by age
data_subset_within$Patient <- factor(data_subset_within$Patient, levels = c("CF185","CF157","CFOP41","CF76","CFOP29","CF205","CF77","CF102","CF178","CFOP42","CF164","CF53","CF90","CF100","CFOP38","CF82","CF179","CF152","CF96","CF147","CF194","CF79","CF109","CF181","CF198","CF182","CF66","CF98","CF169","CF55"))
levels(data_subset_within$Patient)

#Subset data between
data_subset_between<-subset(BAL123_BC, Within_Between=="Between")
View(data_subset_between)

#Figure 3B (stripcharts)
plot1 <- ggplot(data_subset_within, aes(x=Patient, y=BC_value)) + geom_line(position=position_dodge(0.4), col="grey") +  geom_jitter(position=position_dodge(0.4), size=3) + theme(text = element_text(size=20), axis.text.x = element_text(angle = 45, hjust = 1, vjust = 0.9)) + xlab("") + ylab("")

plot2 <- plot1 + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(colour = "black"))	#change theme, no background or gridlines, black axis

plot2

#Figure 3A (boxplots)
#Examine the data distribution for within and between groups
hist(data_subset_within$BC_value, main = "Within", xlab = "Bray-Curtis dissimilarity distance") 
hist(data_subset_between$BC_value, main = "Between", xlab = "Bray-Curtis dissimilarity distance") 

#Neither are normally distributed so look at summary stats and use median and interquartile range as measure of spread
summary(data_subset_within$BC_value)
IQR(data_subset_within$BC_value) 
quantile(data_subset_within$BC_value)

summary(data_subset_between$BC_value) 

IQR(data_subset_between$BC_value) 
quantile(data_subset_between$BC_value) 

#Re-order factors to draw Figure 3A
BAL123_BC$Within_Between <-factor(BAL123_BC$Within_Between, levels=c("Within","Between"))
levels(BAL123_BC$Within_Between)

#Draw boxplots                         
plot3 <- ggplot(BAL123_BC, aes(x=Within_Between, y=BC_value, fill=Within_Between)) + geom_boxplot() + geom_boxplot(outlier.size = 3) + stat_summary(fun=mean, geom="point", shape=4, size=3, color="black") + theme(text = element_text(size=20), axis.text.x = element_text(angle = 45, vjust = 1, hjust=1),legend.position = "none") + scale_fill_manual(values = c("#F8766D", "#619CFF")) + xlab("") + ylab("")

plot4 <- plot3 + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(colour = "black"))	#change theme, no background or gridlines, black axis

plot4
```

```{r, warning=FALSE, error=FALSE, message=FALSE,fig.dim=c(12,8)}
#Put the plots together (Figure 3)
library(cowplot)
#boxplot then stripchart (edited in powerpoint to include labels and patient ages)
plot_grid(plot4,plot2,nrow=1, align = "h", rel_widths = c(1/6,5/6))
```


##Stacked bar charts to visalise microbiota profiles

We will use OTU and metadata tables that include the 4 samples lost during sampling. These samples will be included without any reads in the OTU table (all 0) but can have associated metadata (need to have Patient_age information to produce figures)

```{r, warning=FALSE, error=FALSE, message=FALSE,fig.dim=c(12,10)}
#Read in OTU table and env with the four extra samples in that were lost during subsampling
bacterial_data_4samples<-read.table("Beky_decon_BAL123_IS_4samples_included_OTU_table_R.txt", header=T)
View(bacterial_data_4samples)
meta_table_4samples<-read.csv('Beky_decon_4samples_included_env.csv',check.names=FALSE) 
View(meta_table_4samples)

#Make OTU table into relative abundance
OTU_table_to_format<-bacterial_data_4samples #copy table to format
OTU_total<-cbind(OTU_table_to_format, total = rowSums(OTU_table_to_format)) #add column with rowsums at the end
OTU_order<-OTU_total[order(-OTU_total$total),] #sort according to rowsums
OTU_order_no_total<-subset(OTU_order, select = -c(total)) #subset to remove rowsums column
OTU_order_no_total_matrix<-as.matrix(OTU_order_no_total) #make into matrix
transposed_matrix<-t(OTU_order_no_total_matrix) # transpose
OTU_relabund<-make_relative(transposed_matrix) #calculate relabund (will be proportion between 0 and 1)
transposed_relabund<-t(OTU_relabund) # transpose back the right way
percent_relabund<-(transposed_relabund)*100 # multiply all values by 100 to get %

#Check OTU table has been transformed correctly into relative abundance
View(percent_relabund)

#subset to the Top25 rows (already sorted by rowsums)
Top25_rows<-percent_relabund[1:25,]

#Need to merge Top25 OTU and metadata table

#reformat Top25 OTU
Top25_rows_transformed<-t(Top25_rows) #Transform
View(Top25_rows_transformed)
#make rownames into first column
Top25_rows_transformed_col1names <- cbind(rownames(Top25_rows_transformed), data.frame(Top25_rows_transformed, row.names = NULL)) 
#Rename 'SAMPLE' which is the same column name as the metadata file
names(Top25_rows_transformed_col1names)[1] <- "SAMPLE" 

#merge using common column name 'SAMPLE'
merged_file	<- merge(Top25_rows_transformed_col1names, meta_table_4samples, by.x	= c("SAMPLE"), by.y=c("SAMPLE"))
View(merged_file)

#Change genus names to remove unclassifieds
colnames(merged_file)[colnames(merged_file) == 'Enterobacteriaceae_unclassified'] <- 'Enterobacteriaceae'
colnames(merged_file)[colnames(merged_file) == 'Planctomycetales_unclassified'] <- 'Planctomycetales'

#change data to long format 
library(tidyr)
data_long <- gather(merged_file, Genus, Relabund, Veillonella:Planctomycetales, factor_key=TRUE) 
View(data_long)

#Choose the palette for the stacked bar charts
library('RColorBrewer')
#number of colours required worked out from genus number
colourCount = length(unique(data_long$Genus)) 
#palette selection
getPalette = colorRampPalette(brewer.pal(9, "Set1")) #palette selection

#Reorder genus order so that important pathogens are different colours
data_long$Genus <-factor(data_long$Genus, levels=c("Veillonella","Prevotella_7","Streptococcus","Haemophilus","Selenomonas_3","Prevotella","Pseudomonas","Prevotella_6","Granulicatella","Neisseria","Actinomyces","Alloprevotella","Acinetobacter","Capnocytophaga","Staphylococcus","Leptotrichia","Massilia","Porphyromonas","Fusobacterium","Enterobacteriaceae","Gemella", "Bacillus", "Rothia", "Planctomycetales","Stenotrophomonas"))

#Reorder patient_age into microbiota profile types (the patient_age column has the sample code plus the age in brackets)
data_long$Patient_age<-as.factor(data_long$Patient_age)
data_long$Patient_age <- factor(data_long$Patient_age, levels = c("CF100 (8.5)","CF164 (7.1)","CF181 (14.0)","CF185 (1.1)","CF96 (12.6)","CF98 (16.6)","CFOP41 (1.3)","CFOP42 (6.6)","CF109 (13.8)","CF178 (6.5)","CF169 (16.7)","CF79 (13.2)","CF205 (4.7)","CF157 (1.1)","CF147 (12.6)","CF182 (14.9)","CF76 (2.7)","CF77 (5.2)","CFOP29 (3.1)","CF152 (11.9)","CF53 (7.2)","CF55 (17.7)","CF194 (12.7)","CFOP38 (8.7)","CF102 (5.4)","CF179 (11.8)","CF82 (11.3)","CF66 (15.2)","CF198 (14.8)","CF90 (8.2)"))

#plot by patient_age (Supplementary Figure S3)
#Warning message about 100 rows being removed is due to the 4 patient samples without any read data
#Figure 4 was produced by subsetting different patient numbers for stacked bar charts and combining with tables in powerpoint

plot2<-ggplot(data_long, aes(fill=Genus, y=Relabund, x=Type)) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(colour = "black")) + 
  geom_bar( stat="identity") + scale_fill_manual(values = getPalette(colourCount)) + 
  theme(axis.text.x= element_text(size = 15, colour = "black", angle = 90, vjust = 0.4, hjust = 1)) + xlab("Sample Type") + 
  theme(axis.text.y= element_text(size = 15, colour = "black")) +
  theme(axis.title.x = element_text(size=15, face = "bold", margin = margin(t = 15, r = 0, b = 0, l = 0))) + ylab ("Relative abundance (%)") + 
  theme(axis.title.y = element_text(size=15,face = "bold", margin = margin(t = 0, r = 15, b = 0, l = 0)))

plot2 + facet_wrap(~Patient_age, nrow = 3) + theme(legend.margin = margin(0,0,0,20)) +
  theme(strip.text.x = element_text(size = 15)) + guides(fill=guide_legend(ncol=1)) +
  theme(legend.text=element_text(size=15),legend.title = element_blank(), legend.key = element_rect(size = 2, fill = NA, colour = "white"), legend.key.width = unit(0.8, "cm"),
        legend.key.height = unit(0.8, "cm"))
```


##Checking concordance between bacterial diversity profiles in mock community controls and DNA extraction blanks

As the same methods were applied to check each set of samples (mock community controls and DNA extraction blanks), R scripts are only given for one set (DNA extraction blanks). The stacked bar charts for this analysis were produced using Excel.

```{r, warning=FALSE, error=FALSE, message=FALSE}
#Read in an OTU table with relative abundances for each DNA extraction blank
OTU_table <- read.table("CFSpIT_Beky_decon_blanks_proportions_R.txt", header=T, check.names="FALSE") #Proportions
head(OTU_table)

#The first approach is to check the Pearson Product-moment correlation coefficient (PPMCC) between pairs of blanks 
library(ggpubr)

#Normality testing suggests the data is not normal so PPMCC is not ideal but it has been used previously for this type of data (Muhlebach et al., 2018)
shapiro.test(OTU_table$Maxwell) 
shapiro.test(OTU_table$Maxwell1)
shapiro.test(OTU_table$Maxwell2)

#Blanks are highly correlated with each other
cor.test(OTU_table$Maxwell, OTU_table$Maxwell1, method = "pearson")
cor.test(OTU_table$Maxwell, OTU_table$Maxwell2, method = "pearson")
cor.test(OTU_table$Maxwell1, OTU_table$Maxwell2, method = "pearson")

#The second approach is to check the Bray-Curtis dissimilarity distances between pairs of blanks
#Transpose the OTU table
OTU_table_transposed<-t(OTU_table)

#Calculate BC distances and make into matrix
BC_OTU <-vegdist(OTU_table_transposed, method="bray", binary = FALSE)
matrixOTU=data.matrix(BC_OTU)

#run the flattenCorrMatrix function from earlier on BC matrix
pairwise_flat<-flattenCorrMatrix(matrixOTU)

#The blanks have low Bray-Curtis dissimilarity distances suggested the microbiota profiles are concordant
pairwise_flat
```

